### DISCUSSION TITLE
Assertion failed: (std::this_thread::get_id() == thread.get_id())

### DISCUSSION THREAD

**Discussion by hyn2028** [ANLAB-KAIST/KENSv3#26] [Project Q&A 1. Application Layer (Socket)] [answered]
# Problem
While performing PA1, the following assertion error continues to occur during the test (./echo) using the KENS runtime, and the test is stopped.

`Assertion failed: (std::this_thread::get_id() == thread.get_id()), function run, file E_System.cpp, line 179.`

This error occurs non-deterministically, not when running any particular test.

# Screenshots of Problem
![스크린샷 2022-03-16 오후 8 59 39](https://user-images.githubusercontent.com/5181065/158744922-673802c1-2dea-424d-ab39-6cf46f2ef39b.png)
![스크린샷 2022-03-16 오후 8 59 53](https://user-images.githubusercontent.com/5181065/158744930-4927fe0b-52e1-43d8-a6fd-45fae5f6971c.png)
![스크린샷 2022-03-16 오후 9 08 06](https://user-images.githubusercontent.com/5181065/158744934-6ba44397-f82c-41ea-b78e-35666359bfd5.png)

# Environment
- macOS Monterey 12.3 (x86)
- Apple clang 13.1.6
- cmake 3.22.3

# Extra information
No thread-related functions were used in the implementation of EchoAssignment.cpp, and all functions run within the main thread.

It's probably an assertion that occurred inside the KENS system, so it's difficult for me to debug by myself, and I'm not sure what additional information to provide.

**Comment by Pusnow** [comment]
It might be a memory bug. Follow below instructions to detect which part causes the error.

https://github.com/ANLAB-KAIST/KENSv3/wiki/Tip:-Debugging-Memory-Bugs-(Only-for-macOS-and-Linux)

**Reply by hyn2028** [reply]
I tried to run the program using the address sanitizer, but no memory errors were caught. The `assertion error` keeps popping up. 

This time, I tried to build and run it in a VM used in another class, not in a macOS environment. The environment is:
- Ubuntu 20.04.4 LTS
- cmake version 3.16.3
- gcc (Ubuntu 9.4.0-1ubuntu1~20.04)

In this environment, I ran the `./echo` binary several times, but ***never*** got an `assertion error`. Isn't this a problem that only occurs in a macOS environment?

**Reply by Pusnow** [reply]
It seems macOS-specific problem. I'll check it and fix later.
Thank you for reporting this.

**Comment by Pusnow** [comment, marked as answer]
This is fixed in the latest release (https://github.com/ANLAB-KAIST/KENSv3/releases/tag/v3.3.1). Please, pull latest copy and try agin.

**Reply by hyn2028** [reply]
Thanks for the quick bug fix. As a result of testing in the **macOS** environment, no ```assertion error``` occurs, and all tests pass.
